client/my-sites/themes/themes-selection.jsx

```diff
@@ -191,69 +194,95 @@ function bindGetPremiumThemePrice( state, siteId ) {
-			query,
-			source: sourceSiteId,
-			siteId: siteId,
-			siteSlug: getSiteSlug( state, siteId ),
-			themes: getThemesForQueryIgnoringPage( state, sourceSiteId, query ) || [],
-			themesCount: getThemesFoundForQuery( state, sourceSiteId, query ),
-			isRequesting:
-				isCustomizedThemeListLoading || isRequestingThemesForQuery( state, sourceSiteId, query ),
-			isLastPage: isThemesLastPageForQuery( state, sourceSiteId, query ),
-			isLoggedIn: isUserLoggedIn( state ),
-			isThemeActive: bindIsThemeActive( state, siteId ),
-			isInstallingTheme: bindIsInstallingTheme( state, siteId ),
-			// Note: This component assumes that purchase and plans data is already present in the state tree
-			// (used by the `isPremiumThemeAvailable` selector). That data is provided by the `<QuerySitePurchases />`
-			// and `<QuerySitePlans />` components, respectively. At the time of implementation, neither of them
-			// provides caching, and both are already being rendered by a parent component. So to avoid
-			// redundant AJAX requests, we're not rendering these query components locally.
-			getPremiumThemePrice: bindGetPremiumThemePrice( state, siteId ),
-			filterString: prependThemeFilterKeys( state, query.filter ),
-		};
-	},
-	{ setThemePreviewOptions, recordGoogleEvent, recordTracksEvent }
-)( ThemesSelection );
+			let sourceSiteId;
+			if ( source === 'wpcom' || source === 'wporg' ) {
+				sourceSiteId = source;
+			} else {
+				sourceSiteId = siteId && isJetpack && ! isAtomic ? siteId : 'wpcom';
+			}
+
+			// number calculation is just a hack for Jetpack sites. Jetpack themes endpoint does not paginate the
+			// results and sends all of the themes at once. QueryManager is not expecting such behaviour
+			// and we ended up loosing all of the themes above number 20. Real solution will be pagination on
+			// Jetpack themes endpoint.
+			const number = ! [ 'wpcom', 'wporg' ].includes( sourceSiteId ) ? 2000 : 100;
+			const query = {
+				search,
+				page,
+				tier: premiumThemesEnabled ? tier : 'free',
+				filter: compact( [ filter, vertical ] ).join( ',' ),
+				number,
+			};
+
+			let themes = getThemesForQueryIgnoringPage( state, sourceSiteId, query ) || [];
+
+			// The block templates filter is used to search for full site editing (FSE) themes, which includes both
+			// universal and block themes. If, however, a user is already on a universal theme without FSE enabled
+			// (universal classic), selecting another universal theme will not enable the full site editor. Because
+			// of this, we filter out universal themes from the block-templates filter search results for universal
+			// classic sites. This logic will be removed when FSE is enabled for all sites with universal themes.
+			const currentThemeId = getActiveTheme( state, siteId );
+			const currentTheme = getCanonicalTheme( state, siteId, currentThemeId );
+			const isUniversalClassicSite =
+				! blockEditorSettings?.is_fse_active && currentTheme?.template === 'blockbase';
```

Comment:
I think `is_fse_eligible` might be the condition we want here? As the eligibility sticker is added in the same cases as the site option to enable universal themes for FSE. We also probably dont need the current theme check (and note as in my other comment that blockbase is not a child of blockbase and would need a separate check).

Currently this seems to detects if the user is currently using a universal theme in classic mode, but if they are not currently on a universal theme it gives no indication for whether or not activating a universal theme would result in FSE.

A bug I noticed in testing as a result:
* Take an existing non-FSE enrolled site, active a non-universal block theme like TT2.
* Notice all universal themes now show up under the block templates filter, despite that switching to them would result in classic mode.

Reply:
> Currently this seems to detects if the user is currently using a universal theme in classic mode
Yes! I mistakenly thought that that was our desired outcome.

If I understand things correctly now, we'd want to hide universal themes from _any_ site on the classic experience (which, after giving it more thought, makes sense to me)

One point of confusion for me while making this change was that sites that are `is_fse_eligible: false` _can_ have the site editor enabled 😅 Once we finish the FSE roll out to everyone, which will be soon, I suppose the attribute won't matter, but I just wanted to clarify for any potential readers in the meantime.